tools/tracevis/lib/trace/criticalPath.js

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright 2016 LinkedIn, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/*
+ * Computes critical path in the Parseq plan
+ *
+ * The algorithm performs these steps:
+ *
+ * 1) Traces are translated to execution units. Therefore the original graph never changes. During
+ * this translation we ignore potential parent pointers, potential successor pointers and traces
+ * without start or end timestamp. Graph of execution units records only start and end timestamps
+ * of each trace, predecessor-successor and parent-child relations.
+ *
+ * 2) Each execution unit representing a non-leaf trace (has at least on child) is divided into two
+ * new execution units - front and back. Front represents the time between start of the original
+ * unit and start of its first child, second represents the time between end of its last child and
+ * end of the original unit. Front inherits predecessors, second part successors, both inherit
+ * eventual parent of the original execution unit. Children of the original unit which had no
+ * predecessors, become successors of front, children of the original unit which had no successors,
+ * become predecessors of back. The original unit is deleted. This way we translate all parent and
+ * child pointers to predecessor and successor pointers retaining original timing constaints.
+ *
+ * 3) We find topological sorting of our new graph using this algorithm:
+ * https://en.wikipedia.org/wiki/Topological_sorting#Depth-first_search
+ *
+ * 4) We assign to each task its start offset and end offset. Start offset is the earliest duration
```

Comment:
"duration" typically means time interval, not point in time, so perhaps better word would be "time"?